### PR TITLE
Enables torch.full bool and integer type inference

### PR DESCRIPTION
**Pull request by mruberry**
After being deprecated in 1.5 and throwing a runtime error in 1.6, we can now enable torch.full inferring its dtype when given bool and integer fill values. This PR enables that inference and updates the tests and docs to reflect this.

### CODE REVIEW

File: aten/src/ATen/native/TensorFactories.cpp [hunk 1]
```diff
@@ -354,18 +354,17 @@ TensorOptions infer_full_options(
...
+      return options.dtype(at::kBool);
+    } else if (fill_value.isIntegral(false)) {
+      return options.dtype(at::kLong);
     } else if (fill_value.isComplex()) {
       auto scalar_type = (get_default_dtype() == ScalarType::Double) ?
                             ScalarType::ComplexDouble :
                             ScalarType::ComplexFloat;
       return options.dtype(scalar_type);
+    } else {
+      return options.dtype(get_default_dtype());
```
Comment (gchanan) [inline aten/src/ATen/native/TensorFactories.cpp:367]: do you need to do this?  Presumably it was working before without this?

**Review by gchanan — APPROVED** [review]

File: aten/src/ATen/native/TensorFactories.cpp (hunk 1 above)
Comment (mruberry) [reply aten/src/ATen/native/TensorFactories.cpp:367]: It did and we could safely leave this out. I included it for consistency and readability.

**Comment by albanD** [comment]
@mruberry 
https://github.com/pytorch/pytorch/issues/38011 landed before this one and enabled "warning as error" for the doc. So because this PR introduces new doc warnings, it failed on master.
The failing job is [here](https://app.circleci.com/pipelines/github/pytorch/pytorch/194530/workflows/5286710c-45cd-4018-99db-00d71706db1f/jobs/6417305) and the exact warning is:
```
Jul 29 07:30:45 /opt/conda/lib/python3.6/site-packages/torch/__init__.py:docstring of torch.full:: WARNING: more than one target found for cross-reference 'bool': torch.FloatStorage.bool, torch.Tensor.bool
```

**Comment by mruberry** [comment]
Thanks @albanD! Darn merge conflicts.

**Pull request closed without merging**